File: 2405.1/kotlinchk/security/ServerSideRequestForgeryChecker/ServerSideRequestForgery.kt
package testcase.rule.security.ServerSideRequestForgeryChecker

import java.net.HttpURLConnection
import java.net.URL
import javax.servlet.http.HttpServletRequest

object TestCase {
	fun violation(request: HttpServletRequest) {
		val urlText = request.getParameter("url")
		val url = URL(urlText)
		val connection = url.openConnection() as HttpURLConnection		// @violation
	}

	fun good1(request: HttpServletRequest) {
		val urlWhiteListed = "https://example.com/"
		val urlText = request.getParameter("url")
		if (urlText.startsWith(urlWhiteListed)) {
			val url = URL(urlText)
			val connection = url.openConnection() as HttpURLConnection	// good
		}
	}

	val urlMap: Map<String, URL> = java.util.Collections.emptyMap();

	fun good2(request: HttpServletRequest) {
		val url1 = urlMap.get(request.getParameter("url"))
		val connection1 = url1?.openConnection() as HttpURLConnection	// good

		val url2 = urlMap[request.getParameter("url")]
		val connection2 = url2?.openConnection() as HttpURLConnection	// good
	}
}

File: 2405.1/jschk/security/ServerSideRequestForgery/ServerSideRequestForgery.js
function violation() {
  var http = require('http');
  var url = require('url');

  function listener(request, response) {
    var request_url = url.parse(request.url, true)['query']['url'];     // @violation
    http.request(request_url)
  }
  http.createServer(listener).listen(8080);
}

function good1() {
  var http = require('http');
  var url = require('url');

  function listener(request, response) {
    var urlText = request.url;

    if(urlText == "https://example.com") {                          // good
      var request_url = url.parse(urlText, true)['query']['url'];
      http.request(request_url);
    }
  }
  http.createServer(listener).listen(8080);
}

function good2() {
  var http = require('http');
  var url = require('url');

  function listener(request, response) {
    var urlText = request.url;

    if("https://test.com".equals(urlText)) {                            // good
      var request_url = url.parse(urlText, true)['query']['url']; 
      http.request(request_url);
    }
  }
  http.createServer(listener).listen(8080);
}

function good3() {
  var http = require('http');
  var url = require('url');

  function listener(request, response) {
    var urlText = request.url;

    if(urlText.startsWith("http://")) {                                 // good
      var request_url = url.parse(urlText, true)['query']['url']; 
      http.request(request_url)
    }
  }
  http.createServer(listener).listen(8080);
}